src/core/ddsi/src/ddsi_wraddrset.c

```diff
@@ -784,7 +787,7 @@ struct addrset *compute_writer_addrset (const struct writer *wr, bool enable_shm
   }
   else
   {
-    struct costmap *wm = wras_calc_costmap (covered, prefer_multicast, enable_shm);
+    struct costmap *wm = wras_calc_costmap (covered, prefer_multicast, (wr->xqos->present & QP_LOCATOR_MASK) ? wr->xqos->ignore_locator_type : 0);
```

Comment:
With the locator mask always present in the QoS, there's no need for the conditional anymore. (I personally would probably add an assert: those things verify invariants/assumptions and usually reduce the amount of state I have to track in my mind, instead of enlarging it like this does.)

Reply:
Maybe the assert can be put inside the validate_reader_qos function?
Then that function does what it says on the tin? :)

Reply:
🙂

But no, not really. `validate_reader_qos` is about validating input, this is about verifying that some invariant (in this case, reader/writer->xqos actually has the locator mask initialized) holds. If you check every invariant everywhere, it becomes really easy to write flawless software, the only thing is that you never get it done 🙂

In this case, I think it would be useful at point of use, because things will go horrible wrong in hard to detect ways if it is uninitialized, and if such a check exists in a few key places any screw ups will get caught quickly.

Reply:
I agree with your point about not overchecking things, but I personally do not like using asserts to validate input, while they are fine in debug builds, they are a little too unreliable for release builds in my eyes.

Reply:
For general input validation, they are pretty bad, no doubt about that. For checking assumptions and invariants, they are (unfortunately) the best we have in C, because a real run-time check means you get a system that behaves differently from what you expect. In the end that really makes it harder to reason about which states are valid, and which aren't. In my experience, in the long run, that means you give up and you simply *assume* that certain states are valid because it is tested for somewhere and you end up propagating the assumption to other places until eventually no-one remembers that, actually, that state is impossible. (That was really a painful lesson to learn ...)

I can live with not checking at all (not even assert — but that assertion is essentially free, so why not?), but I do have a problem with handling the `!(present & QP_LOCATOR_MASK)` case (here), when it cannot be because it is included in the default set.

If that makes any sense 🙂